### DISCUSSION TITLE
HMR

### DISCUSSION THREAD

**Discussion by ryanflorence** [remix-run/remix#2384] [Proposals]
We thought we wanted it in the beginning, but we're not sure now. With the full-stack nature of Remix it's quite nice to have the full page reload so you don't break your app in production.

For example, if you do this:

```js
let [state, setState] = useState(0)
```

Fire up your app, start working and then change it to this:

```js
let [state, setState] = useState(localStorage.foo)
```

With HMR, your app continues to work just fine. If you don't hit "refresh" and try to server render that code, you'll never know you broke your app.

Our current live reload approach ensures that every change runs through a full server render pass so you don't end up with those kinds of surprises.

Additionally, when you're changing loader code, HMR won't update the UI, live reload does.

Finally, usually HMR is more about hiding the slowness of front-end build tools than about maintaining client state on code changes. ESBuild is fast, so it doesn't matter.

All that said, we've got some experimental stuff working from a while back, but it's really not high priority right now because there are some tradeoffs to consider, it's not just an obvious thing to add. Also the live reload experience right now is so good that some people think we already have HMR 🤫

This issue is here, and left open, because we haven't decided to do HMR, but we haven't decided not to either.

EDIT: 03/22

I'm not sure why I didn't mention it when we created this discussion cause it's been on our minds the whole time 🙃 but the biggest argument that keeps me believing we should do the work to add this is that it's *super annoying* to have to wait for your loaders when you're just screwing around with the padding on some elements. My workflow in Remix is like 10% JSX/loaders/actions/react stuff and 90% fiddling with styles. That 90% would be a lot nicer with HMR.

**Comment by alexturpin** [comment]
I appreciate the Remix team leaving the door open to this, and so just wanted to weigh in for team HMR a bit!

I understand Remix's goals with SSR first, and I also agree that with Edge computing, it's likely where everything is headed. However, I assume there are many others like me who are building apps with significant client-side state and interaction in order to create really rich UX. One example of such an app would be an online collaborative editor, something like Figma or Miro.

It would be awesome to have some fast refresh capabilities to make building that kind of apps on Remix an even nicer experience. On my current project, I even use the [Vite HMR API](https://vitejs.dev/guide/api-hmr.html) to enhance the development experience according to our business requirements.

**Comment by itsMapleLeaf** [comment]
I've worked on a handful of "app-like apps" using Remix lately, and this is still near the top of my wish list.

Although ESBuild runs are fast, in my opinion, maintaining state between changes is a really underrated capability. It helps in cases like editing a popover state while it's open, editing a menu item while it's open, toggles, and so on. The failure case as described in the OP is a valid one for sure. But I don't know if guarding against that is worth sidestepping a potentially huge boost in DX.

**Comment by leon** [comment]
My story is I want to use remix as the framework for a react-three-fiber app.

I really like remix, but compared to using vite and their react plugin I get real time updates both to css and 3d code which is a must have once you have tried it.

Could we tell remix which folders to do fast refresh on?

that way we could get fast refresh on ui components and in my case the react-three-fiber app, but still have full reloads when changing loaders or route adjacent modules.

**Comment by dan-lee** [comment]
I think it's vital for an ambitious project like Remix to provide HMR/ fast refresh as powerful as vite. I understand the argument that a full page reload can help prevent bugs but most of the time changes don't warrant a full page reload.
I'd love to see ongoing investigation how this could be achieved

**Comment by fishuke** [comment]
I believe you can provide it as an option if possible. Like two seperate scripts:

```bash
npm run dev
npm run dev:hmr
```

**Comment by agcty** [comment]
Remix absolutely needs HMR, it's an amazing framework and I really wanted to migrate from Next.js but not having HMR is quite a dealbreaker for us, unfortunately.

We are building a game with react-three-fiber and have a lot of 3d models that are being loaded into a canvas.

A Canvas component could look like this:



```jsx
const models = [model1, model2, ....., model100]

<Canvas>
    <ambientLight />
    <pointLight position={[10, 10, 10]} />
    <Suspense fallback={null}>
        {models.map((x) => (
          <MyModel src={x} />
        ))}
    </Suspense>
</Canvas>
```

each model could have a size anywhere between 50kb - 1mb in our case.

Without HMR, Remix becomes literally unusable as rebuilding the canvas can take anywhere from 20s to up to **1 minute** 🥲

I understand it's non-trivial to make HMR work adequately client and server-side and that reloading in a lot of cases with LiveReload is enough, but there are a few cases where HMR becomes absolutely crucial.

**Reply by gopeter** [reply]
Literally the same for our app. Fairly large internal project management app which is built on top of dozens of components from our internal design system. Changing the style of a popup drives you crazy if the state is lost on every change.

**Reply by tiagoad** [reply]
For data visualization this would also be very useful (for the same reasons)

**Comment by agcty** [comment]
Is there anything I could help with getting HMR out of the way? Would love to know what the current issues are and what code/files I should take a look at. As @kentcdodds mentioned in discord, CSS is probably the most important thing that benefits from HMR in the short term.

**Comment by aulneau** [comment]
I am also very interested in this -- i am finally moving toward exploring remix as an option compared to next.js or vite based apps, and HMR is a huge loss for me here. I'd be happy to help contribute whatever I can towards this, I think many, many people would benefit from this. It sounds like there are a lot of good ideas too: opt in, making it clear etc etc. @ryanflorence would love if you could update us on if this is something that is coming 🙏

**Comment by shogs** [comment]
Can we have some sort of official update on this and if it is a priority now or not. I was very keen on using Remix on a new project but with the lack of HMR support makes it not so ideal now. However, if it's coming in soon I may put up with Live Reloading for the time being until HMR gets in.

Some insight, my upcoming project will need a lot of custom styling and modelling with heavy data requirement so live reloading wouldn't be ideal, though for now mocking a subset of the data would do. I've also used HMR quiet a bit in the past and I find it great for styling and design changes without reloading the entire page, sure at times you need to reload but thats never been a problem and I have never deploying without a final check or test of my app.

Thanks.

**Reply by kiliman** [reply]
The team has been busy working on some major functionality lately: moving Remix data handling to React Router, adding support for streaming data and deferred. Not to mention they are hosting Remix Conf next week.

HMR is on their radar, but it's hard to say when it'll land.

**Reply by aulneau** [reply]
Had the team considered a move towards using vite in the backend? Would give hmr and much more benefits imo

**Reply by itsMapleLeaf** [reply]
Ryan has expressed passing interest on Twitter: https://twitter.com/ryanflorence/status/1520059699109920773

**Reply by kentcdodds** [reply]
We can't promise when this will be supported. We're definitely thinking about it, we're just not working on it actively as we have other more impactful things we're working on.

We think you'll more than make up for the time "lost" by not having HMR by having the simpler mental model and onboarding process that Remix offers :)

**Reply by aulneau** [reply]
> We think you'll more than make up for the time "lost" by not having HMR by having the simpler mental model and onboarding process that Remix offers :)

I know that you mean it in the kindest way possible, but this kind of statement is pretty dismissive and condescending to a huge group of folks who are genuinely expressing a real need for something that has important use cases (regardless if you happen to share in them). Just food for thought @kentcdodds

**Reply by kentcdodds** [reply]
Definitely didn't mean to be dismissive. I just want to emphasize that when deciding which framework to use, in addition to HMR support people should factor in the other developer productivity gains that Remix has to offer before dismissing it from the get-go.

**Reply by aulneau** [reply]
Appreciate the clarification ✨

**Reply by kiliman** [reply]
Hmm.. you're more than welcome to try implementing HMR if you really need it. As Kent says, it's on their TODO list, but not a high priority. I'm not sure how that's being dismissive or condescending. Just because a lot of people may want it doesn't mean that resources to build it just magically appear.

I think a more constructive response is, "Thanks for considering it. Is there anything I can do to help facilitate that?"

**Reply by kentcdodds** [reply]
@kiliman I think @aulneau was providing feedback on how my comment made them and possibly others feel which I appreciate :)

**Reply by aulneau** [reply]
> Hmm.. you're more than welcome to try implementing HMR if you really need it. As Kent says, it's on their TODO list, but not a high priority. I'm not sure how that's being dismissive or condescending. Just because a lot of people may want it doesn't mean that resources to build it just magically appear.
> 
> I think a more constructive response is, "Thanks for considering it. Is there anything I can do to help facilitate that?"

Thanks @kiliman ! Yeah as @kentcdodds mentioned it was more around all the comments around the usefulness of a given feature. 

I'd absolutely be happy to help contribute where I can and others have as well, but there's been no real comment around that from the team, and PRs implementing changes have been closed with no comments from the team 🤷‍♂️ 

Not trying to start any arguments or anything, just wanted to provide some pointed feedback around perception and narrative. 

Appreciate the work y'all are doing :)

**Reply by itsMapleLeaf** [reply]
> We can't promise when this will be supported. We're definitely thinking about it, we're just not working on it actively as we have other more impactful things we're working on.
> 
> We think you'll more than make up for the time "lost" by not having HMR by having the simpler mental model and onboarding process that Remix offers :)



> Definitely didn't mean to be dismissive. I just want to emphasize that when deciding which framework to use, in addition to HMR support people should factor in the other developer productivity gains that Remix has to offer before dismissing it from the get-go.

I agree that Remix offers a ton to reduce overall development time. That being said, I think HMR is less about objective time saves, and more about making for a nicer, smoother experience. Keeping state in place while making changes that show up immediately, that definitely feels a lot smoother than full page refreshes :slightly_smiling_face:

**Reply by itsMapleLeaf** [reply]
On the subject of "be the change you want to see", I've actually played around a bit with the idea of slotting in Vite to Remix's compiler, but this is not a simple problem to solve :sweat_smile: I ended up moving on to other things after that. Not to say it's hopeless; I'm no compiler expert :joy:

**Reply by shogs** [reply]
Thanks for the responses, sorry I didn't get a chance to respond sooner. I've actually decided to use Remix because of other factors but with the hopes that HMR will eventually come, maybe with the migration to Vite if it ever happens. Thanks...

**Reply by tigerabrodi** [reply]
@itsMapleLeaf Donny, creator of SWC told me he knew nothing about compilers when starting with SWC 😂 

nothing is impossible 🔥

**Reply by JClackett** [reply]
Literally praying for HMR support! I have a 1 route app, which utilises a lot of modals and every time I save I have to open the modal. Defaulting the modal to open is the temp fix but not ideal :)

**Reply by hollandThomas** [reply]
@JClackett Have you considered putting each modal in a nested route and navigating there instead of setting some state to open it? I'm all for HMR as well but in that particular case, this might just be the solution

https://github.com/remix-run/remix/tree/main/examples/route-modal

**Reply by JClackett** [reply]
@hollandThomas yeah just switched everything to modals and this helps a lot! Thank you - still definitely a lot of use cases for hmr :)

**Comment by eagsalazar** [comment]
Any updates?  Also I feel like the outlined issues with HMR aren't particularly specific to Remix and yet everyone wants it and benefits from it, so sounds like people mostly are ok with the limitations.

**Comment by hkan** [comment]
I would also suggest implementing the feature into core Remix libraries but having it disabled by default, and documenting it clearly. This way Remix authors achieve the default setup without the aforementioned pitfalls, but developers have the freedom to accept the risk and have HMR.

If anyone is keeping record of choices, I'm also pro-HMR.

**Reply by JClackett** [reply]
Agreed! No HMR is great when everything you are doing can be referenced from the URL, but there's so many use cases where this isn't suitable.

**Comment by karibertils** [comment]
Trying Remix for the first time. Looks promising, but coming from Next.js the lack of HMR feels like a huge step backwards for me. HMR is far from perfect on Next.js and very prone to breaking things, I add `// @refresh reset` to top of files all the time to force full reload when things start acting up. And yet I feel it's absolutely a must have killer feature.

**Comment by bradchristensen** [comment]
I just wanted to add that if you're coming here looking for a faster alternative to Next.js that supports HMR, I recommend giving Vite and [vite-plugin-ssr](https://vite-plugin-ssr.com/) a go. It's probably only worth looking at if your project is currently large and slow to build/develop though, rather than a little hobby project you want to get up and running quickly. By that I mean it will likely require more investment to get up and running - and bear in mind that it won't give you the same ergonomics that you get with Remix out of the box. But if, like me, all you're after is a _significant_ speed boost coming from Next.js, that'll do the trick. As an example, my app would typically take ~5-10s to load each page in Next.js, but page loads with vite-plugin-ssr are each <1s :fire:

(And by the way, I love Remix and really hope to see HMR support at some point in the future!)

**Reply by lu-zen** [reply]
Vite is awesome.

**Comment by ourmaninamsterdam** [comment]
This is a great debate. I love Remix and I'd love to see HMR added.

We're currently evaluating tech stacks and are down to a choice between Next.js and Remix. DX being one of our core decision-making factors. So without the lack of HMR we're leaning towards Next.js. If we knew it was on the roadmap then it could swing us back.

Related to this, when you make a change to a component in Next.js, any compilation errors are piped into the browser and displayed.  Whereas Remix renders the previously compiled page and shows the compilation error in the CLI, so you often miss that an error has occurred. Is there a way around this?

**Reply by hkan** [reply]
> Remix renders the previously compiled page and shows the compilation error in the CLI

That's not the case for me. If there is an error it shows the error on the page. Share this issue in the Discord community's help channel, maybe?

**Reply by cliffordfajardo** [reply]
+1 to what hkan said ^

> Remix renders the previously compiled page and shows the compilation error in the CLI
If there is an error in the server, I see it show in the browser - its similar to nextjs experience

**Reply by kiliman** [reply]
Not sure if OP is the same person, but someone on Discord had a similar issue. Turns out he was using the Express adapter and `const build = require(BUILD_DIR)` at the top, instead of dynamically in the request handler.

Anyway, Remix should work out the box. Please share a simple repro if you're still having issues.

**Reply by ourmaninamsterdam** [reply]
Runtime errors get shown in the browser without issue, but build errors aren't shown, only in the command-line. IMO it's not a serious issue, but I've been caught out a few times where I've continued making changes (non-UI) and hadn't realised that it wasn't rebuilding.

Example repo https://github.com/ourmaninamsterdam/remix-error-test

It appears there's another discussion about build errors happening here: https://github.com/remix-run/remix/discussions/4196

**Reply by kiliman** [reply]
Yeah, I should publish my patch that I used in #4196

**Comment by SimeonGriggs** [comment]
It it were possible and made the feature simpler to ship: I'd be in favour of per-route, opt-in HMR.

In my instance I have an SPA (like a client-side-only dashboard) loaded at a specific route which would benefit greatly from HMR, the rest of the application is fine without it.

**Reply by machour** [reply]
Would love to have a Remix debug bar, with the ability to tick/untick HMR whenever I want it

**Comment by sudomf** [comment]
Hey all. I just started [sudomf/remix-vite](https://github.com/sudomf/remix-vite). It's in the very beginning. But I think I'll be able to support most of remix apps asap.

Give it a try by running.

```sh
npx remix-vite
```

**Reply by pcattori** [reply]
@sudomf and I chatted today about how his `remix-vite` dev server is implemented. Mayke, correct me if I miss anything.

## Technical summary

[`vite.ts`](https://github.com/sudomf/remix-vite/blob/main/src/vite.ts) is a dev server using [Vite SSR](https://vitejs.dev/guide/ssr.html#setting-up-the-dev-server).[^prod]

[^prod]: The Vite server is not used in production, just in development.
 
There are two plugins that are Rollup ports of the esbuild plugins we use in the Remix compiler:
- [`remix`](https://github.com/sudomf/remix-vite/blob/main/src/plugins/remix.ts): provides Remix virtual modules
- [`transform`](https://github.com/sudomf/remix-vite/blob/main/src/plugins/transform.ts): treeshakes route modules for browser and server

There are two more plugins to enable Vite + HMR:
- [`inject`](https://github.com/sudomf/remix-vite/blob/main/src/plugins/inject.ts): Replaces `<Scripts />` with the `@vite/client` runtime and replaces `<LiveReload />` with React Fast Refresh entrypoint
- [`hmr-fix`](https://github.com/sudomf/remix-vite/blob/main/src/plugins/hmr-fix.ts): Adapts React Fast Refresh to Remix route modules by relaxing the assumption that a module only exports a component

Finally, you can run `remix-vite` as a CLI or you can the [Node API to create Express-compatible middleware](https://github.com/sudomf/remix-vite#custom-remix-server) for the custom Vite server.

## Limitations

Currently, this implementation only works for React (since it relies on React Fast Refresh) and only for the Node server runtime (so not for Cloudflare nor Deno).

## Proposal

Continue exposing `remix dev` for live reload without HMR.
Expose `remix dev:hmr` and a middleware API for opt-in HMR (limited to Node + React). 

For non-Node runtimes and non-React renderers, continue using `remix dev` for live reload until we can add support for those.

**Reply by OnurGvnc** [reply]
> Limitations
Currently, this implementation only works for React (since it relies on React Fast Refresh) and only for the Node server runtime (so not for Cloudflare nor Deno).

Removing backend-only exports for frontend at [transform.ts:45](https://github.com/sudomf/remix-vite/blob/e6a5389e8d25f7c01115ac9cad892c5cfcfbb427/src/plugins/transform.ts#L45)
There is no transform for the backend/SSR at [transform.ts:25](https://github.com/sudomf/remix-vite/blob/e6a5389e8d25f7c01115ac9cad892c5cfcfbb427/src/plugins/transform.ts#L25)

Let's say we ran the remix dev server normally. (node or deno or cloudflare)

If 'action' and 'loader' functions for backend/ssr in remix-vite/src/plugins/transform.ts are transformed with functions like the below, the actual backend code will work on the runtime we want.

```js
export async function loader({ request }) {
  const url = new URL(request.url)
  const response = await fetch(
    new Request(
      `http://localhost:${process.env.ORIGINAL_REMIX_DEV_SERVER_PORT}${url.pathname}?_data=${PATH}`,
      request
    )
  )
  const contentType = response.headers.get('content-type')
  if (contentType && contentType.indexOf('application/json') !== -1) {
    const data = await response.json()
    return json(data)
  }
  // other implementations...
}
```

In this scenario, remix dev server and remix-vite are running at the same time. this may be considered as a disadvantage, but we make sure that the 'loader' and 'action' methods work correctly in the correct runtime.

**Reply by sudomf** [reply]
Hey @pcattori. Yes, it's all correct.

**Reply by sudomf** [reply]
@OnurGvnc I think it can work as the `remix dev:hmr` idea is meant just to be used for development. The users must be warned that when they run remix with hmr, they will be running a code that differs from production. And I think it's ok when we realize the existent solutions for hmr, like Vite, Webpack, already run one version of the code for dev and another version for production.

**Reply by hiyelbaz** [reply]
![image](https://github.com/remix-run/remix/assets/1831733/5ae0ecaf-265e-4a2a-9298-3228f63f87e4)

Should we just ignore it? @pcattori

Here is the related part of .eslintrc.cjs
![image](https://github.com/remix-run/remix/assets/1831733/9b597e24-c856-421d-a1bd-05a1dbcf0322)

**Reply by pcattori** [reply]
Yes, you can safely ignore this as Remix handles Remix-specific exports like `meta`, `handle`, etc... for you

**Comment by acusti** [comment]
i’m super excited by v1.14.0 and the new development server with HMR (and Hot Data Revalidation)! this looks exactly like what i’ve been hoping for. however, i am using the cloudflare-workers adapter for remix, so my `yarn dev` run script runs `remix watch` + `miniflare` simultaneously as the dev server, and i don’t use the remix dev server (no `remix dev`). can i simply replace my existing `"dev:remix": "remix watch --debug"` npm run script with `"dev:remix": "remix dev"` in my package.json to enable this functionality? in other words, does the remix dev server play nicely with miniflare?

**Reply by acusti** [reply]
in browsing through issues mentioning “HMR”, i found this: https://github.com/remix-run/remix/issues/4838

the proposal in that issue mentions that the new remix dev server “handles the client assets dynamically” while “other requests are proxied to the real server”, and that “`remix.config.js` has a new `dev` field that indicates how to spawn your development server”. so it seems like i could have a `"dev:remix": "remix dev"` npm run script plus `{ "dev": "cross-env NODE_ENV=development miniflare ./build/index.js --watch --port 3000 --open" }` in my `remix.config.js` to use the new dev server + HMR. is that the right way to do it?

**Reply by acusti** [reply]
this is getting a little silly, but i keep finding new resources. from #4838 (the issue about the new dev server), i saw that it was released in v1.12, which led me to the accompanying [release notes](https://github.com/remix-run/remix/releases/tag/remix%401.12.0). from those notes:

> At this stage, the idea is to spin up the dev server alongside your normal Remix app server:
> ```
> # spin up the new dev server
> remix dev
> 
> # Spin up your app server in parallel.
> # This can be done in a separate terminal or with a tool like `concurrently`.
> nodemon --watch build/ ./server.js
> ```

so the answer to my original question, which was “can i simply replace `remix watch` → `remix dev`”, seems to be “yes” (once i’ve added   `future: { unstable_dev: true },` to my `remix.config.js`). if i have issues, i will report back, but otherwise, i’m very excited to give this a go. thanks again to everyone who worked on these updates!

**Reply by acusti** [reply]
i couldn’t get it to work. here’s what i tried: i upgraded all remix packages to v1.14.0, added `future: { unstable_dev: true },`, replaced my existing `remix watch` run script with `cross-env NODE_ENV=development remix dev`, and removed the `LiveReload` component from `app/root.tsx`. CSS file updates result in a line that looks like `GET /__REMIX_ASSETS_MANIFEST 200 OK (1.27ms)` in my dev script output, but no update to the page in the browser. same with TSX component updates.

**Reply by lili21** [reply]
Yeah. Need a better doc about how to enable it.

**Reply by dmarkow** [reply]
@acusti Why did you remove the `LiveReload` component? That's what's needed for the HMR updates to work.

**Reply by acusti** [reply]
@dmarkow i removed it because i first tried the upgrade / migration with the `LiveReload` component still in place and it didn’t work, plus i started getting lots of websocket errors in the browser console, so i figured the new dev server was a full replacement for the previous `LiveReload` component.

it makes sense that it would still be required, so i tried restoring it and then tried a bunch of different configs to get the websocket server working and communicating with the client (e.g. no LiveReload `props.port` and no remix.config.js `future.port`, `<LiveReload port={8002} />` and remix.config.js `future: { port: 8002 }`, `<LiveReload port={3010} />` and remix.config.js `devServerPort: 3010, { future: { port: 3010 } }`, all of those config options plus `cross-env NODE_ENV=development REMIX_DEV_SERVER_WS_PORT=8002 remix dev` in the run script, etc.), but i still just get `[Error] WebSocket connection to 'ws://localhost:8002/socket' failed: Could not connect to the server.` (where 8002 is whatever i configured as the port) in the browser console and no HMR.

**Reply by jgentes** [reply]
I'm in the same boat, tried many options but just get a reloading page due to websocket unable to connect to 8002. Just going without `unstable_dev` for now - it's unstable!
